### PR TITLE
fix(gnovm): forbid star expression when value is not a pointer

### PR DESCRIPTION
**Pull request by omarsy**
closes: #1088 

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
</details>

### CODE REVIEW

**Comment by notJoon** [comment]
It looks fine to me.  However, I found that for the following case, the error message slightly differs from what Go outputs:

```go
package main

const VALUE = 5

func main() {
    println(*VALUE)
}
```

When this code is executed in Go, it produces the following error message:

```plain
invalid operation: cannot indirect VALUE (untyped int constant 5)
```

However, when run this code in gno, you can see `const 5` is output instead of the constant name `VALUE`.

```plain
panic: invalid operation: cannot indirect (const (5 <untyped> bigint)) (variable of type <untyped> bigint) [recovered]
        panic: main/gnovm/tests/files/ptr10.gno:6:13: invalid operation: cannot indirect (const (5 <untyped> bigint)) (variable of type <untyped> bigint):
        --- preprocess stack ---
        stack 2: func main() { (const (println func(xs ...interface{})()))(*((const (5 <untyped> bigint)))) }
        stack 1: file{ package main; const VALUE<VPBlock(2,0)> = (const (5 <untyped> bigint)); func main() { (const (println func(xs ...interface{})()))(*((const (5 <untyped> bigint)))) } }
        stack 0: package(main)
        ------------------------
```

I think it would be helpful for debugging if it were made to output the constant's name like Go does.

Additionally, if you declare it with `var` instrade of `const`, `VALUE` is output correctly.

```go
package main

var VALUE = 5

func main() {
    println(*VALUE)
}

// Output: panic: invalid operation: cannot indirect VALUE<VPBlock(3,0)> (variable of type int) [recovered]
```

**Comment by omarsy** [comment]
> It looks fine to me. However, I found that for the following case, the error message slightly differs from what Go outputs:
> 
> ```go
> package main
> 
> const VALUE = 5
> 
> func main() {
>     println(*VALUE)
> }
> ```
> 
> When this code is executed in Go, it produces the following error message:
> 
> ```
> invalid operation: cannot indirect VALUE (untyped int constant 5)
> ```
> 
> However, when run this code in gno, you can see `const 5` is output instead of the constant name `VALUE`.
> 
> ```
> panic: invalid operation: cannot indirect (const (5 <untyped> bigint)) (variable of type <untyped> bigint) [recovered]
>         panic: main/gnovm/tests/files/ptr10.gno:6:13: invalid operation: cannot indirect (const (5 <untyped> bigint)) (variable of type <untyped> bigint):
>         --- preprocess stack ---
>         stack 2: func main() { (const (println func(xs ...interface{})()))(*((const (5 <untyped> bigint)))) }
>         stack 1: file{ package main; const VALUE<VPBlock(2,0)> = (const (5 <untyped> bigint)); func main() { (const (println func(xs ...interface{})()))(*((const (5 <untyped> bigint)))) } }
>         stack 0: package(main)
>         ------------------------
> ```
> 
> I think it would be helpful for debugging if it were made to output the constant's name like Go does.
> 
> Additionally, if you declare it with `var` instrade of `const`, `VALUE` is output correctly.
> 
> ```go
> package main
> 
> var VALUE = 5
> 
> func main() {
>     println(*VALUE)
> }
> 
> // Output: panic: invalid operation: cannot indirect VALUE<VPBlock(3,0)> (variable of type int) [recovered]
> ```

@notJoon 
Thank you for your input! The difference in how `ConstExpr` and `NameExpr` are displayed stems from the specific formatting logic defined in their respective implementations. You can see this in the `nodes_string.go` file:

- For `ConstExpr`, the display logic is found [here](https://github.com/gnolang/gno/blob/master/gnovm/pkg/gnolang/nodes_string.go#L552).
- For `NameExpr`, the relevant logic is located [here](https://github.com/gnolang/gno/blob/master/gnovm/pkg/gnolang/nodes_string.go#L98).

**Comment by notJoon** [comment]
> > It looks fine to me. However, I found that for the following case, the error message slightly differs from what Go outputs:
> > ```go
> > package main
> > 
> > const VALUE = 5
> > 
> > func main() {
> >     println(*VALUE)
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When this code is executed in Go, it produces the following error message:
> > ```
> > invalid operation: cannot indirect VALUE (untyped int constant 5)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, when run this code in gno, you can see `const 5` is output instead of the constant name `VALUE`.
> > ```
> > panic: invalid operation: cannot indirect (const (5 <untyped> bigint)) (variable of type <untyped> bigint) [recovered]
> >         panic: main/gnovm/tests/files/ptr10.gno:6:13: invalid operation: cannot indirect (const (5 <untyped> bigint)) (variable of type <untyped> bigint):
> >         --- preprocess stack ---
> >         stack 2: func main() { (const (println func(xs ...interface{})()))(*((const (5 <untyped> bigint)))) }
> >         stack 1: file{ package main; const VALUE<VPBlock(2,0)> = (const (5 <untyped> bigint)); func main() { (const (println func(xs ...interface{})()))(*((const (5 <untyped> bigint)))) } }
> >         stack 0: package(main)
> >         ------------------------
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think it would be helpful for debugging if it were made to output the constant's name like Go does.
> > Additionally, if you declare it with `var` instrade of `const`, `VALUE` is output correctly.
> > ```go
> > package main
> > 
> > var VALUE = 5
> > 
> > func main() {
> >     println(*VALUE)
> > }
> > 
> > // Output: panic: invalid operation: cannot indirect VALUE<VPBlock(3,0)> (variable of type int) [recovered]
> > ```
> 
> @notJoon Thank you for your input! The difference in how `ConstExpr` and `NameExpr` are displayed stems from the specific formatting logic defined in their respective implementations. You can see this in the `nodes_string.go` file:
> 
>     * For `ConstExpr`, the display logic is found [here](https://github.com/gnolang/gno/blob/master/gnovm/pkg/gnolang/nodes_string.go#L552).
> 
>     * For `NameExpr`, the relevant logic is located [here](https://github.com/gnolang/gno/blob/master/gnovm/pkg/gnolang/nodes_string.go#L98).

Thanks for response. it seems reasonable to me. So I'll remove the triage-pending flag.

**Review by notJoon — APPROVED** [review]

**Review by thehowl — APPROVED** [review]
LGTM.

File: gnovm/pkg/gnolang/preprocess.go [hunk 1]
```diff
@@ -1738,7 +1738,12 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 			case *KeyValueExpr:
 				// NOTE: For simplicity we just
 				// use the *CompositeLitExpr.
-
+			// TRANS_LEAVE -----------------------
+			case *StarExpr:
+				xt := evalStaticTypeOf(store, last, n.X)
+				if xt.Kind() != PointerKind && xt.Kind() != TypeKind {
+					panic(fmt.Sprintf("invalid operation: cannot indirect %s (variable of type %s)", n.X.String(), xt.String()))
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:1745]: consider this one:
```go
package main

func main() {
	println(*nil)
}
```
but yes I'll make a new PR.

Comment (thehowl) [reply gnovm/pkg/gnolang/preprocess.go:1745]: always ahead of the game :) should have let you take a look first

Comment (omarsy) [reply gnovm/pkg/gnolang/preprocess.go:1745]: nice catch. Missed this one ^^

Comment (omarsy) [reply gnovm/pkg/gnolang/preprocess.go:1745]: did a fix here: https://github.com/gnolang/gno/pull/3053

**Pull request merged**